pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,463 @@
+			}
+		}
+
+		gvk, err := apiDes.GetGroupVersionKind()
+		if err != nil {
+			return err
+		}
+		if _, ok := s.apiDescriptionMap[gvk]; ok {
+			return fmt.Errorf("descriptors with duplicated GVK %v are found", gvk)
+		}
+		api := apiDes
+		s.apiDescriptionMap[gvk] = &api
+	}
+
+	return nil
+}
+
+// appendTostructuredObjectMap converts the unstructured object and try to convert it into a specified interface
+// before appending to the unstructuredObjectMap.
+func (s *synthesize) appendTostructuredObjectMap(kind string, bufObj unstructured.Unstructured) {
+	s.unstructuredObjectMap[kind] = append(s.unstructuredObjectMap[kind], bufObj)
+}
+
+// loadDependencies tries to decode an object into registry.DependenciesFile{}.
+// It updates the dependencies in synthesize once decode is successful.
+func (s *synthesize) loadDependencies(path string) error {
+	dependenciesFile := registry.DependenciesFile{}
+	if err := registry.ParseDependenciesFile(path, &dependenciesFile); err != nil {
+		if strings.HasPrefix(err.Error(), "Unable to decode the dependencies file") {
+			return nil
+		}
+		return err
+	}
+
+	s.dependencies = &dependenciesFile
+	return nil
+}
+
+// getNonInferableCSV gets the NonInferableCSV object ensures that is loaded.
+func (s *synthesize) getNonInferableCSV() (*NonInferableCSV, error) {
+	if s.nonInferableCSV.Name != "" {
+		return s.nonInferableCSV, nil
+	}
+	return nil, CsvInferenceFileNotFoundError
+}
+
+// getInstallStrategy converts all Deployment kind object in bundle into install strategy and extracts the
+// service account name referred in the deployments. It then adds all roleRefs from (cluster)RoleBindings that mentions
+// any of the service accounts, adding the rules of (cluster)roles paring with their correspondent ServceAccounts into
+// (cluster)permissions.
+func (s *synthesize) getInstallStrategy() (*operatorsv1alpha1.NamedInstallStrategy, error) {
+	deploys, err := s.getDeployments()
+	if err != nil {
+		return nil, fmt.Errorf("deployments are required for synthesizing CSV, %v", err)
+	}
+
+	var deploymentSpecs []operatorsv1alpha1.StrategyDeploymentSpec
+	deploySANames := make(map[string]struct{})
+
+	for _, deploy := range deploys {
+		deployName := deploy.Name
+		if deployName == "" {
+			deployName = names.SimpleNameGenerator.GenerateName(deploy.GetGenerateName() + "-")
+		}
+
+		deploymentSpecs = append(deploymentSpecs, operatorsv1alpha1.StrategyDeploymentSpec{Name: deployName, Spec: deploy.Spec})
+
+		dname := deploy.Spec.Template.Spec.ServiceAccountName
+		if dname == "" {
+			dname = "default"
+		}
+		deploySANames[dname] = struct{}{}
+	}
+
+	// Get all RBAC w/ subjects matching Deployment ServiceAccounts by name only.
+	var permissions []operatorsv1alpha1.StrategyDeploymentPermissions
+	rolesRefAndSANameMap := make(map[rv1.RoleRef]string)
+	rbs, err := s.getRoleBindings()
+	if err != nil && err != ObjectTypeNotFoundInBundleError {
```

Comment:
This logic could be simplified throughout if these resource getters returned nil slices rather than a sentinel error (which would allow synthesis to rely on the behavior of range over empty slices instead).